src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+        if (other) |o| {
+            if (self.num_spaces == o.num_spaces) {
+                return self.num_tabs > o.num_tabs;
+            } else if (self.num_tabs == o.num_tabs) {
+                return self.num_spaces > o.num_spaces;
+            } else {
+                return null;
+            }
+        } else {
+            // No minimum indent means we’re always “more indented.”
+            return true;
+        }
+    }
+};
+
+/// A simple line record: the index in the token list and its indent.
+pub const Line = struct {
+    token_start: u32,
+    indent: Indent,
+};
+
+/// The buffer that accumulates tokens.
+pub const TokenizedBuffer = struct {
+    allocator: std.mem.Allocator,
+    tokens: std.MultiArrayList(Token),
+    lines: std.ArrayList(Line),
+
+    pub fn init(allocator: std.mem.Allocator) !TokenizedBuffer {
+        return TokenizedBuffer{
+            .allocator = allocator,
+            .tokens = std.MultiArrayList(Token){},
+            .lines = std.ArrayList(Line).init(allocator),
+        };
+    }
+
+    /// Pushes a token with the given kind, token offset, and length.
+    pub fn pushToken(self: *TokenizedBuffer, kind: T, tok_offset: usize, tok_length: usize) !void {
+        try self.tokens.append(self.allocator, .{
+            .kind = kind,
+            .offset = @intCast(tok_offset),
+            .length = @intCast(tok_length),
+        });
+    }
+
+    /// Returns the offset of the token at index `idx`.
+    pub fn offset(self: *TokenizedBuffer, idx: u32) usize {
+        return @intCast(self.offsets.items[@intCast(idx)]);
+    }
+};
+
+/// A comment is represented by its begin and end offsets.
+pub const Comment = struct {
+    begin: usize,
+    end: usize,
+};
+
+/// MessageKind enumerates different diagnostic messages.
+pub const MessageKind = enum {
+    MisplacedCarriageReturn,
+    AsciiControl,
+    LeadingZero,
+    UnknownToken,
+    OpaqueNameWithoutName,
+    UppercaseBase,
+    InvalidUnicodeEscapeSequence,
+    InvalidEscapeSequence,
+    UnclosedString,
+    UnclosedSingleQuote,
+    BadNumberSuffix,
+    OverClosedBrace,
+    MismatchedBrace,
+};
+
+const UnicodeKind = enum {
+    LetterUpper,
+    LetterNotUpper,
+    Digit,
+    Other,
+    Invalid,
+};
+
+const UnicodeInfo = struct {
+    kind: UnicodeKind,
+    length: usize,
+};
+
+/// A message with its kind and offset.
+pub const Message = struct {
+    kind: MessageKind,
+    begin: u32,
+    end: u32,
+};
+
+/// The cursor is our current position in the input text, and it collects messages.
+/// Note that instead of allocating its own message list, the caller must pass in a pre-allocated
+/// slice of Message. The field `message_count` tracks how many messages have been written.
+/// This can grow beyond the length of the slice, and if so, it means there are more messages
+/// than the caller has allocated space for. The caller can either ignore these messages or
+/// allocate a larger slice and tokenize again.
```

Comment:
Why? Why not just use an arraylist of messages?

Reply:
* This avoids `try` in a few places (no allocation necessary - therefore it doesn't need to be fallible!)
* Realistically, we don't want an arbitrary-length list here. Past 100 messages, you're likely getting gunk. If your file is that broken, it's probably not roc.

Reply:
SG!